docs/TDD.md

```diff
@@ -0,0 +1,220 @@
+# Test-Driven Development
+
+Test-driven development ([TDD][TDD]) refers to a style of programming where tests are written as artifacts to guide the implementation of the program design in code.
+
+One or more tests (particularly [unit tests][unit tests]) are written before coding.
+The tests are intended to cover one aspect of the program's behavior, which may be focused on a single function or method.
+The writing of the tests is a way to transform the program requirements and general architecture into an implementation-specific design.
+The tests are run, and they should fail, because the code hasn't been implemented.
+The code is implemented and the tests are run again.
+If the tests pass, then either implementing the behavior is done or perhaps more necessary tests remain to be created.
+If the tests don't pass, then the code is debugged and the tests are run again.
+The cycle of testing and coding is repeated until all of the necessary tests pass, at which point the implementation for that aspect of the program's behavior is done... for now.
+
+## Refactoring
+
+[Refactoring][refactoring] is the rewriting of code to improve its design.
+It is not simply rewriting code to fix bugs.
+It is sometimes called "refactoring" to modify code to get it to pass the tests.
+Although modifying the code _may_ include improving the design as a way to pass the tests, simply debugging is not necessarily improving the _design_ of the code, and so is not necessarily _refactoring_.
+
+Following is an example of debugging without refactoring:
+
+```python
+
+# A function intended to return x added to y.
+# x and y are bad parameter names, but we ignore that for now.
+def add(x, y):
+    # used multiply operator by mistake. It fails the tests.
+    return x * y
+
+# Function corrected. It passes the tests. It has been debugged, but not refactored.
+def add(x, y):
+    return x + y
+    
```

Comment:
```suggestion

```

Reply:
Can't commit this as it became outdated after another commit.

Reply:
That is fine, you can always make the change and attribute.  Github does nothing that git will not do.